libs/api/ebpf_api.cpp

```diff
@@ -4920,18 +4929,123 @@ CATCH_NO_MEMORY_EBPF_RESULT
+                map_handle,
+                reinterpret_cast<ebpf_handle_t>(GetCurrentProcess()),
+                &local_subscription->perf_event_array_map_handle,
+                0,
+                FALSE,
+                DUPLICATE_SAME_ACCESS)) {
+            result = win32_error_code_to_ebpf_result(GetLastError());
+            _Analysis_assume_(result != EBPF_SUCCESS);
+            EBPF_LOG_WIN32_API_FAILURE(EBPF_TRACELOG_KEYWORD_API, DuplicateHandle);
+            EBPF_RETURN_RESULT(result);
+        }
+
+        // Get user-mode address to perf buffer shared data.
+        ebpf_operation_map_query_buffer_request_t query_buffer_request{
+            sizeof(query_buffer_request),
+            ebpf_operation_id_t::EBPF_OPERATION_MAP_QUERY_BUFFER,
+            local_subscription->perf_event_array_map_handle,
+            cpu_id};
+        ebpf_operation_map_query_buffer_reply_t query_buffer_reply{};
+        result = win32_error_code_to_ebpf_result(invoke_ioctl(query_buffer_request, query_buffer_reply));
+        if (result != EBPF_SUCCESS) {
+            EBPF_RETURN_RESULT(result);
+        }
+        ebpf_assert(query_buffer_reply.header.id == ebpf_operation_id_t::EBPF_OPERATION_MAP_QUERY_BUFFER);
+        local_subscription->buffer =
+            reinterpret_cast<uint8_t*>(static_cast<uintptr_t>(query_buffer_reply.buffer_address));
+
+        // Initialize the async IOCTL operation.
+        local_subscription->callback_context = callback_context;
+        local_subscription->sample_callback = sample_callback;
+        local_subscription->lost_callback = lost_callback;
+        local_subscription->cpu_id = cpu_id;
+        memset(&local_subscription->reply, 0, sizeof(ebpf_operation_map_async_query_reply_t));
+        result = initialize_async_ioctl_operation(
+            local_subscription.get(),
+            _ebpf_perf_event_array_map_async_query_completion,
+            &local_subscription->async_ioctl_completion);
+        if (result != EBPF_SUCCESS) {
+            EBPF_RETURN_RESULT(result);
+        }
+
+        // Issue the async query IOCTL.
+        ebpf_operation_map_async_query_request_t async_query_request{
+            sizeof(async_query_request),
+            ebpf_operation_id_t::EBPF_OPERATION_MAP_ASYNC_QUERY,
+            local_subscription->perf_event_array_map_handle,
+            cpu_id,
+            query_buffer_reply.consumer_offset};
+        result = win32_error_code_to_ebpf_result(invoke_ioctl(
+            async_query_request,
+            local_subscription->reply,
+            get_async_ioctl_operation_overlapped(local_subscription->async_ioctl_completion)));
+        if (result != EBPF_SUCCESS) {
+            if (result == EBPF_PENDING) {
+                result = EBPF_SUCCESS;
+            } else {
+                local_subscription->async_ioctl_failed = true;
+            }
+        }
+
+        // If the async IOCTL failed, then free the subscription object.
+        if (result == EBPF_SUCCESS) {
```

Comment:
this commend reads slightly confusing to me - you mention failure, but this check if for EBPF_SUCCESS case.